onnxruntime/core/optimizer/constant_sharing.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef ENABLE_TRAINING
+
+#include <limits>
+
+#include "core/framework/op_kernel.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/constant_sharing.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+/**
+ * @brief Share initializer for those who hold same value in same type and shape.
+ *
+ * @tparam T The initializer value type, used to retrieved data values.
+ * @param graph Target graph to iterate.
+ * @param node Target node to check initializer input.
+ * @param input_index Input index of target node.
+ * @param type_value_plus_rank_to_shared_arg_map Accumulated map from value/rank to initializer.
+ *  The key indiciate initializer's data type, value and rank.
+ *  The value is the first initializer NodeArg* to be shared.
+ * @param data_type Initializer data type.
+ * @return true If the input initializer can be shared.
+ * @return false If the input initializer CANNOT be shared.
+ */
+template <typename T>
+bool ConstantSharing::ShareInitializer(Graph& graph, Node* node, int input_index,
+                                       std::map<std::string, NodeArg*>&
+                                           type_value_plus_rank_to_shared_arg_map,
+                                       int32_t data_type) const {
+  InlinedVector<T> values;
+  const bool require_constant = true;
+  const NodeArg* input_def = node->InputDefs()[input_index];
+  if (!optimizer_utils::AppendTensorFromInitializer(graph, *input_def, values, require_constant) ||
+      values.size() != 1) {
+    return false;
+  }
+
+  int rank = input_def->Shape()->dim_size();
+  std::ostringstream oss;
+  oss << data_type << "_" << values[0] << "_" << rank;
+  std::string p = oss.str();
+  if (type_value_plus_rank_to_shared_arg_map.find(p) == type_value_plus_rank_to_shared_arg_map.end()) {
+    type_value_plus_rank_to_shared_arg_map[p] = const_cast<NodeArg*>(input_def);
+    return false;
+  } else {
+    graph_utils::ReplaceNodeInput(*node, input_index, *type_value_plus_rank_to_shared_arg_map[p]);
+    graph.RemoveConsumerNode(input_def->Name(), node);
+    if (graph.GetConsumerNodes(input_def->Name()).size() == 0) {
```

Comment:
Do we need to add a consumer node for the initializer that is being used as the replacement?

Reply:
You are right!!! Good catch! thanks a lot!!